December 10 2024 - adding James Nord spotlight (#227)
jenkins-infra/contributor-spotlight: src/pages/contributors/james-nord.adoc

== What is your background prior to contributing to Jenkins?

After receiving my master's in electrical and electronic engineering from Manchester University, I moved to Sweden and started life as a research engineer at a university.
I was working with location-aware applications and mass distribution of media on the internet.
This was all before mobile phones had built-in GPS systems.
After some time in Sweden, I moved to Hampshire in the UK to work on the "head end" (backend) for a company catering to satellite and cable TV operators.
Within broadcasting, they call it "head end" and it's everything that drives automation of the broadcast.
This included EPGs, scheduling, and encryption of data and devices as well.
So I was working on the management of life cycles, encryption for devices, and other various necessities.

> The company was acquired by Cisco partway through my ten-year tenure there, which is when I got my first exposure to Hudson.

Reviewer: something snot quite right here.
I think this implies I got the exposure at Cisco, but it was with the company before it was acquired that I started using it.

Reviewer: Looks good, thanks

While we weren't _technically_ doing them from developer desktops, we were doing releases from servers, but you would log into the server and do it from a server, which is just as bad.
That led to my first plugin, which was the link:https://plugins.jenkins.io/m2release/[M2 Release (Maven Release)] plugin.
I wrote that so we could do our releases from Hudson, and then it evolved from there.
Later on, there were some instances where we started using Cucumber for automated tests.
Since there was no good way to show Cucumber test results in Jenkins, I wrote the link:https://plugins.jenkins.io/cucumber-reports/[Cucumber reports] plugin.
From there, my usage was mostly running and maintaining Hudson.
I filed loads of bugs, while also fixing an odd bug here and there.

During my time at CloudBees, I have focused on Enterprise Jenkins or CloudBees CI and all of its various parts.
10 years ago, when I started to work for CloudBees, I was working on the proprietary side of CloudBees core.
So I was working on a lot of plugins for CloudBees and sometimes having to fix bugs in open source.
I've also attempted to maintain the acceptance test harness.

> Most recently, I've been driving the FIPS effort for our proprietary products, which means we have had to have FIPS hooks inside Jenkins.

Reviewer: not sure if this is worth a link too, do people know what FIPS 140 is?

```suggestion
Most recently, I've been driving the link:https://csrc.nist.gov/pubs/fips/140-2/upd2/final[FIPS 140] effort for our proprietary products, which means we have had to have FIPS 140 hooks inside Jenkins.
```

> Most recently, I've been driving the FIPS effort for our proprietary products, which means we have had to have FIPS hooks inside Jenkins.

Author: I think it's worth linking to here, thanks for catching that @jtnord

We've been working with the team and managing that to get enough into open source so that we can build on it.

> There's a JEP and various plugins that we have contributed towards the FIPS effort.

Reviewer: or link to the specific one?
```suggestion
There's a link:https://github.com/jenkinsci/jep/tree/master/jep/237[JEP] and various plugins that we have contributed towards the FIPS effort.
```

> There's a JEP and various plugins that we have contributed towards the FIPS effort.

Author: Definitely, thanks for sharing that @jtnord

> CloudBees CI has now gone GA with support for FIPS 140, so the base work for that is obviously in the open source.

Author: That's a great callout, thanks @jtnord

However, I was the lab manager for all the manual testing and we had racks of servers that were no longer powerful enough to do encryption.
Since they were laying around, essentially being spare parts, I cobbled together pieces from various old servers to give it more memory and then shoved it back in the rack.
There was no budget for it and no IT for it, but it was in the rack, ready to go.
It was brilliant and we loved it, with other people eventually catching on.
They saw what we were doing and they wanted that ability too.
So, along with my regular job of developing code and maintaining the lab, I basically took on owning CI that eventually snowballed and spread out across the division.
We were a global company at the time, so it eventually became a question of "How do we make this a bit more global?"
Since we had source repositories in different places, we had Jenkins (Hudson) in three different environments; two in the UK and one in Israel so it would always be close to the source code.

== How long have you been using Jenkins?

So, if you take when Hudson was first made available and add six months onto it, that would probably be about the timeline.

> After leaving my previous employer, I moved over to CloudBees and I've been working for ClubBees for the last nine and a half, almost 10 years.

Author: oh gosh I missed that one, thanks @gounthar

During my time at CloudBees, I have focused on Enterprise Jenkins or CloudBees CI and all of its various parts.
10 years ago, when I started to work for CloudBees, I was working on the proprietary side of CloudBees core.
So I was working on a lot of plugins for CloudBees and sometimes having to fix bugs in open source.
I've also attempted to maintain the acceptance test harness.

Most recently, I've been driving the FIPS effort for our proprietary products, which means we have had to have FIPS hooks inside Jenkins.
We've been working with the team and managing that to get enough into open source so that we can build on it.
There's a JEP and various plugins that we have contributed towards the FIPS effort.
CloudBees CI has now gone GA with support for FIPS 140, so the base work for that is obviously in the open source.
However, it's not as easy to run and you need to do your own checks.

== Why choose Jenkins over other projects?

While I was working for NDS, the satellite cable company, as we were scaling up and out, we wanted to be able to allow users to create their own projects.
However, we wanted them to essentially be rubber stamp templates.

> So it's like this: you're always going to use a Maven project, so the only thing that is different is where your source control is located and what email should be notified when something goes wrong.

Reviewer 2: ```suggestion
So it's like this: you're always going to use a Maven project, so the only different thing is where your source control is located and what email should be notified when something goes wrong.
```

That was around the time that Kohsuke had just started his company InfraDNA.

> I believe the first time that I worked with Kohsuke, was the Nectar VMware autoscaling plugin which allowed us to have a pool of resources in VMware that were just virtual machines and just resetting to snapshots and assign them into builds.

Reviewer 2: ```suggestion
I believe the first time that I worked with Kohsuke, was on the Nectar VMware autoscaling plugin which allowed us to have a pool of resources in VMware that were just virtual machines and just resetting to snapshots and assigning them into builds.
```

> So we could say, for this build I need a virtual machine; here you go, here's an IP address, and we were good to go.

Reviewer 2: ```suggestion
So we could say, for this build, I need a virtual machine; here you go, here's an IP address, and we were good to go.
```

> I was amazed at the quickness in which he knocked out that plugin.

Reviewer 2: ```suggestion
I was amazed at the quickness with which he knocked out that plugin.
```

While in Jerusalem, we mentioned something during a meeting and he came in the next day and said here you go, try this.
It was almost out of nowhere and the collaboration with Kohsuke was amazing.

> Additionally, at the time it had a cutting edge UI and the test result reports were great.

Reviewer 2: ```suggestion
Additionally, at the time it had a cutting-edge UI and the test result reports were great.
```

If you compared it to other stuff, it was a little table of pass/fail or a mess.
Hudson allowed you to have a summary, drill into things, and show you the full stack trace and full logs right at the bottom of the builds, as opposed to having all the information in a massive sprawling page of mess.

== What problems have you solved in Jenkins?

The most recent thing is being able to support FIPS 140 for any federal people that want to use Jenkins and do it themselves.
Up until a year ago, you couldn't really start Jenkins in a FIPS compliant way because the initial use password that it generates would be generated in something that was not an approved encryption standard.

Another one of the things I did recently, which if you end up having to support Jenkins, is all of the plugin class loaders are now named.

> If something does go wrong and you get a stack trace, rather than having just the class, you have plugin name and class so you can immediately see which plugin is having an issue.

Reviewer 2: ```suggestion
If something does go wrong and you get a stack trace, rather than having just the class, you have the plugin name and class so you can immediately see which plugin is having an issue.
```
